fs/localfs/local_fs.go

```diff
@@ -223,13 +234,25 @@ type fileWithMetadata struct {
 	*os.File
 }
 
+func newFileEntry(fi os.FileInfo, parentDir string) fs.Entry {
```

Comment:
i think this will need to be moved to OS-specific file, since on Windows there's no `*syscall.Stat_t` at all.

OTOH, on linux/darwin you should be able to assume `fi.Sys()` is always `*syscall.Stat_t`

Reply:
It's your codebase so you can have it how you want it, but the presence or absence of an inode is not a dependency on the platform but on the filesystem. It is in principle possible for the libraries provided on windows to allow reading/writing ext4 and for go to implement *syscall.Stat_t in that case and this code will work on all platforms where the filesystem has the property of being based on inodes and the libraries provided by go are written to correctly report it.

On NTFS, the File ID is pretty much analogous to the inode number and there is an argument to be made that go should support reporting that as the inode and then this would even benefit filesystems where it's not, strictly speaking, an inode number yet serves the same function and has similar enough semantics. if `syscall.Stat_t` used `GetFileInformationByHandle` on Windows, I believe the behaviour would be OK.

So in my view, this implementation is the correct way from `kopia`'s perspective, and then it is up to the go standard library authors to decide what makes sense re: `syscall.Stat_t` in their view.

Reply:
I was just making a comment that this won't compile on Windows, because `syscall.Stat_t` does not exist there. Look at the `Build / Make (windows-latest)` check. That's how Golang does it - `syscall` package is GOOS-specific.

I think NTFS FileID makes sense to use in this context, it's just getting that information requires different code on Windows than on Linux.

Tip: you can try cross-compiling for Windows `GOOS=windows GOARCH=amd64 go build`

Reply:
Oh! Well I think that's quite wrong of go but that's another matter. Thanks for the tip on cross-compiling, I can see that you're right and it doesn't build with those settings. 

I've not the slightest idea what the setup is here for os-specific code. Looks like there are files named _windows.go and _linux.go, but no such examples in the fs subtree. I'll need a clue as to how to do it so that it fits with the intended build system(s).

Reply:
There are already windows and non-windows specific files for localfs. Same directory. In golang naming conventions matter, which is probably controversial but works quite well.

Reply:
Doh! I ran `find fs -name '*linux*'` instead of just looking at all the files, and failed to find anything for linux. But I see now it's named `_nonwindows.go`. I'll rewrite this later tonight or tomorrow.

Reply:
OK this I think is an OK fix to the platform problem. Now where should we discuss building the cache?

I would like to do a similar thing to what I've done to local_fs to repofs.go. That is, if the repofs is a snapshot of a filesystem whose entries `HasInode`, then the repofs version of it should also `HasInode`. This will be ignored in any case, as references to the inodes in snapshot uploads will be guarded by a flag that's default off. 

However based on our other discussion I assume you don't like that approach...

Reply:
This is great so far.

I don't understand the use of storing inodes in repository, we have a file identifier already `oid` which is globally unique. If we want to preserve hardlinks on restore, we can do the reverse process there - build a map of `oid` to inodes of restored files and if we find `oid` reuse we do a hard-link to a previously-written file.

As for building the cache, there are different ways of doing this - I would keep some sort of LRU cache with bounded size that's configurable with a default of zero, but somebody might want to set it to a high value like 10M. The cache/map would be `(dev,inode)-> *snapshot.DirEntry` if we find a match, we double check metadata (size, mtime) and if so, reuse the dir entry.

Reply:
The use of having the inodes in the repo dirEntry is that we can have a cache hit before hashing anything, just like with filenames. So if the user does:
```
$ mv bigtree bigtreeNewHome
```
and snapshots, the snapshot will be instant due to device/inode/metadata hit, as though the pathnames hadn't all changed. The inodes are not useful for restore, and I guess I don't know what the `oid` is -- does it correspond to something in my filesystem?

In my particular case, it's also that I want to run a snapshot on one of my already backed up time machine directories with the goal of populating the inode information quickly. On a small example, this is the difference bettween the initial snapshot hashing everything which takes 90 minutes and having all cache hits which takes 10 minutes. I haven't timed it on my large example.

Reply:
[is oid the internal name of the hash? Then that is not useful for avoiding hashing a newly found file at upload time, while a stored inode is]

Reply:
Also, presumably whatever mechanism solves the cache size problem for filenames also solves it for inodes. That is, if the entire repository snapshot is not loaded into ram, but only the `dirEntry`s that correspond to the currently uploading directory, in a similar way only the `dirEntry` whose inode matches the currently uploading `dirEntry` inode needs to be in memory and so there's no need for a new caching mechanism, just another cache like the existing prevEntries sorted by inode instead of filename. I do not really understand hwo the existing prevEntries cache works, but it would be subject to exactly the same problems as the proposed inode-based cache.

Reply:
I don't think I understand how this works. Want to hop on slack so we can chat interactively?

Reply:
I think what you're looking for is some lightweight machine-specific inode->oid cache. prevEntries will not suffice here, because it's only for the current directory, so would not handle anything but trivial renames in one dir.

When it comes to memory usage, we only keep in-memory prevEntries for the directories that are currently "on the stack" during upload, which avoids bloat.